bindings/wasm/src/did/wasm_resolved_document.rs

```diff
@@ -0,0 +1,121 @@
+impl WasmResolvedDocument {
+  /// Attempts to merge changes from a `DiffMessage` into this document and
+  /// updates the `ResolvedDocument::diffMessageId`.
+  ///
+  /// If merging fails the document remains unmodified, otherwise this represents
+  /// the merged document state.
+  ///
+  /// See `Document::mergeDiff`.
+  ///
+  /// # Errors
+  ///
+  /// Fails if the merge operation or signature verification on the diff fails.
+  #[wasm_bindgen(js_name = "mergeDiffMessage")]
+  pub fn merge_diff_message(&mut self, diff_message: &WasmDiffMessage) -> Result<()> {
+    self.0.merge_diff_message(&diff_message.0).wasm_result()?;
+    Ok(())
+  }
+
+  // ===========================================================================
+  // Properties
+  // ===========================================================================
+
+  /// Returns the inner DID document.
+  ///
+  /// NOTE: clones the data. Use `intoDocument()` for efficiency.
+  #[wasm_bindgen(getter)]
+  pub fn document(&self) -> WasmDocument {
+    WasmDocument::from(self.0.document.clone())
+  }
+
+  /// Consumes this object and returns the inner DID document.
+  ///
+  /// NOTE: trying to use the `ResolvedDocument` after calling this will throw an error.
+  #[wasm_bindgen(js_name = intoDocument)]
+  pub fn into_document(self) -> WasmDocument {
+    WasmDocument::from(self.0.document)
+  }
+
+  /// Returns the diff chain message id.
+  #[wasm_bindgen(getter = diffMessageId)]
+  pub fn diff_message_id(&self) -> String {
+    self.0.diff_message_id.to_string()
+  }
+
+  /// Sets the diff chain message id.
+  #[wasm_bindgen(setter = diffMessageId)]
+  pub fn set_diff_message_id(&mut self, value: &str) -> Result<()> {
+    let message_id: MessageId = MessageId::from_str(value).wasm_result()?;
+    self.0.diff_message_id = message_id;
+    Ok(())
+  }
+
+  /// Returns the integration chain message id.
+  #[wasm_bindgen(getter = metadataPreviousMessageId)]
+  pub fn integration_message_id(&self) -> String {
+    self.0.integration_message_id.to_string()
+  }
+
+  /// Sets the integration chain message id.
+  #[wasm_bindgen(setter = integrationMessageId)]
+  pub fn set_integration_message_id(&mut self, value: &str) -> Result<()> {
+    let message_id: MessageId = MessageId::from_str(value).wasm_result()?;
+    self.0.integration_message_id = message_id;
+    Ok(())
+  }
+
+  // ===========================================================================
+  // JSON
+  // ===========================================================================
+
+  /// Serializes a `Document` object as a JSON object.
+  #[wasm_bindgen(js_name = toJSON)]
+  pub fn to_json(&self) -> Result<JsValue> {
+    JsValue::from_serde(&self.0).wasm_result()
+  }
+
+  /// Deserializes a `Document` object from a JSON object.
+  #[wasm_bindgen(js_name = fromJSON)]
+  pub fn from_json(json: &JsValue) -> Result<WasmResolvedDocument> {
+    json.into_serde().map(Self).wasm_result()
+  }
```

Comment:
I think this function and `WasmDocument::from_json` allow creating invalid documents, for example, without any verification methods. Is that intended? Shouldn't these functions use `IotaDocument::verify_document`? I think this would also apply to other `from_json` functions in the bindings.

Reply:
I'm unsure since it should still go through the serde validation functions (i.e. `try_from` and `deserialize_with`). An `IotaDocument` without any verification methods is technically still valid, just deactivated as far as I know but that's beside the point.

I'll check and get back to you.

Reply:
Looks like serde validates it. So as long as there are deserialisation checks in Rust the Wasm bindings will also reject invalid deserialisations.

E.g. trying to decode the following document without a `publicKeyMultibase` field in the verification method:

```js
let document = Document.fromJSON({
        d: {
            id: 'did:iota:2gy3XVL5qP6BngW2Avain8jQpnEVurAPzaAx2q64NaiJ',
            capabilityInvocation: [{
                id:"did:iota:2gy3XVL5qP6BngW2Avain8jQpnEVurAPzaAx2q64NaiJ#sign-0",
                controller:"did:iota:2gy3XVL5qP6BngW2Avain8jQpnEVurAPzaAx2q64NaiJ",
                type:"Ed25519VerificationKey2018",
            }]
        },
        m: {
            created: '2021-12-13T15:00:05Z',
            updated: '2021-12-13T15:00:05Z',
        }
    });
```

throws an error:

```
Error [serde_json::Error]: data did not match any variant of untagged enum MethodRef at line 1 column 266
```

Reply:
It also rejects an invalid `IotaDocument` (wrong DID method id), so it is being validated correctly.
Error:
```
Error [serde_json::Error]: Invalid Method Name at line 1 column 339
```
Example:
```
let document = Document.fromJSON({
        d: {
            id: 'did:invalid:2gy3XVL5qP6BngW2Avain8jQpnEVurAPzaAx2q64NaiJ',
            capabilityInvocation: [{
                id:"did:iota:2gy3XVL5qP6BngW2Avain8jQpnEVurAPzaAx2q64NaiJ#sign-0",
                controller:"did:iota:2gy3XVL5qP6BngW2Avain8jQpnEVurAPzaAx2q64NaiJ",
                type:"Ed25519VerificationKey2018",
                publicKeyMultibase:"zA1WDcdxGnmBxzYR6CB2KJN4iyP7NxPmbZU1BxLXbDd7C"
            }]
        },
        m: {
            created: '2021-12-13T15:00:05Z',
            updated: '2021-12-13T15:00:05Z',
            proof: {
                type: 'JcsEd25519Signature2020',
                verificationMethod: '#sign-0',
                signatureValue: '5kCvWaLpTki2wLt78efDKnQe3W37RZizULN4UY6yUWNRnoHGqJkvQ7Kw54mGZd1pjYvjmYAv8sa7DUFsgyJFnwQw'
            }
        }
    });
```

Reply:
Thanks for checking. This doesn't throw for me, even though it has no method at all:

```js
Document.fromJSON({
        d: { id: 'did:iota:DtwycXocVMpbdgjq6rqXxanCNnv6rAHrYwypKB9RST7h' },
        m: { created: '2021-12-13T15:27:26Z', updated: '2021-12-13T15:27:26Z' }
    });
```

Reply:
Oh yeah, of course :man_facepalming: Nevermind.